{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aa1996a7-7832-484b-8fe3-8e39e9cf83ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Classes reloading.\n",
      "functions reloading\n"
     ]
    }
   ],
   "source": [
    "# RUN\n",
    "import sys\n",
    "sys.path.append(\"/opt/src\")\n",
    "import mip_functions as mip\n",
    "import calculate_prevalences as cap\n",
    "import probe_summary_generator\n",
    "import pickle\n",
    "import json\n",
    "import copy\n",
    "import math\n",
    "import os\n",
    "import numpy as np\n",
    "import subprocess\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('ggplot')\n",
    "from matplotlib.lines import Line2D\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import plotly.express as px\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "import allel\n",
    "wdir = \"/opt/analysis/\"\n",
    "data_dir = \"/opt/data/\"\n",
    "import plotly.express as px\n",
    "from ipywidgets import interact, interactive, fixed, interact_manual\n",
    "import ipywidgets as widgets\n",
    "from urllib.request import urlopen"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b220bad",
   "metadata": {},
   "source": [
    "# Calling genotypes, prevalences and filtering data\n",
    "The original vcf file created by freebayes contain the genotypes determined by the program itself. In addition, genotype values for aggregated and non-aggregated nucleotides and aminoacids are also available as *_genotypes_table.csv files as described above.  \n",
    "\n",
    "However, the default parameters generating the vcf file are not very strict. In this part of the analysis we will apply various filters to the count tables and generate genotype calls based on those filters."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aeb239b1",
   "metadata": {},
   "source": [
    "### Chose which tables to analyse\n",
    "Select the type of data to analyse. Make sure the count file is matching the coverage file. e.g. alternate_XX_table and coverage_XX_table, XX must be the same value (AA, AN or nothing)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83ba304e",
   "metadata": {},
   "source": [
    "#### Example cell\n",
    "```python\n",
    "mutation_count_file = \"/opt/analysis/alternate_AA_table.csv\"\n",
    "mutation_coverage_file = \"/opt/analysis/coverage_AA_table.csv\"\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a903a821",
   "metadata": {},
   "outputs": [],
   "source": [
    "# USER INPUT\n",
    "\n",
    "mutation_count_file = \"alternate_AA_table.csv\"\n",
    "mutation_coverage_file = \"coverage_AA_table.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8f26ed6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Gene ID</th>\n",
       "      <th colspan=\"6\" halign=\"left\">PF3D7_0417200</th>\n",
       "      <th colspan=\"4\" halign=\"left\">PF3D7_0523000</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PF3D7_1408000</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PF3D7_1408100</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th colspan=\"6\" halign=\"left\">dhfr-ts</th>\n",
       "      <th colspan=\"4\" halign=\"left\">mdr1</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PM2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PM3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation Name</th>\n",
       "      <th>dhfr-ts-Ala16Val</th>\n",
       "      <th>dhfr-ts-Asn51Ile</th>\n",
       "      <th>dhfr-ts-Cys59Arg</th>\n",
       "      <th>dhfr-ts-Ile164Leu</th>\n",
       "      <th>dhfr-ts-Ser108Asn</th>\n",
       "      <th>dhfr-ts-Ser108Thr</th>\n",
       "      <th>mdr1-Ala821Thr</th>\n",
       "      <th>mdr1-Asn1042Asp</th>\n",
       "      <th>mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>mdr1-Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>PM2-Asp440Asn</th>\n",
       "      <th>PM2-Gln442His</th>\n",
       "      <th>PM2-Ile401Leu</th>\n",
       "      <th>PM2-Leu277Val</th>\n",
       "      <th>PM2-Leu321Val</th>\n",
       "      <th>PM2-Leu398Val</th>\n",
       "      <th>PM2-Phe350Val</th>\n",
       "      <th>PM2-Thr289Ile</th>\n",
       "      <th>PM3-Arg31Lys</th>\n",
       "      <th>PM3-Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ExonicFunc</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>...</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA Change</th>\n",
       "      <th>Ala16Val</th>\n",
       "      <th>Asn51Ile</th>\n",
       "      <th>Cys59Arg</th>\n",
       "      <th>Ile164Leu</th>\n",
       "      <th>Ser108Asn</th>\n",
       "      <th>Ser108Thr</th>\n",
       "      <th>Ala821Thr</th>\n",
       "      <th>Asn1042Asp</th>\n",
       "      <th>Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>Asp440Asn</th>\n",
       "      <th>Gln442His</th>\n",
       "      <th>Ile401Leu</th>\n",
       "      <th>Leu277Val</th>\n",
       "      <th>Leu321Val</th>\n",
       "      <th>Leu398Val</th>\n",
       "      <th>Phe350Val</th>\n",
       "      <th>Thr289Ile</th>\n",
       "      <th>Arg31Lys</th>\n",
       "      <th>Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Targeted</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>...</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KB-04-06-PRX-04-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-00-45-PRX-00-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KB-07-13-PRX-07-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>243.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-04-08-PRX-04-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-07-84-PRX-07-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 133 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene ID              PF3D7_0417200                                    \\\n",
       "Gene                       dhfr-ts                                     \n",
       "Mutation Name     dhfr-ts-Ala16Val dhfr-ts-Asn51Ile dhfr-ts-Cys59Arg   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                 Ala16Val         Asn51Ile         Cys59Arg   \n",
       "Targeted                       Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1              0.0              2.0              2.0   \n",
       "AG-00-45-PRX-00-1              0.0             30.0             30.0   \n",
       "KB-07-13-PRX-07-1              0.0            254.0            254.0   \n",
       "AM-04-08-PRX-04-1              0.0             18.0             18.0   \n",
       "AM-07-84-PRX-07-1              0.0             51.0             51.0   \n",
       "\n",
       "Gene ID                                                                  \\\n",
       "Gene                                                                      \n",
       "Mutation Name     dhfr-ts-Ile164Leu dhfr-ts-Ser108Asn dhfr-ts-Ser108Thr   \n",
       "ExonicFunc         missense_variant  missense_variant  missense_variant   \n",
       "AA Change                 Ile164Leu         Ser108Asn         Ser108Thr   \n",
       "Targeted                        Yes               Yes               Yes   \n",
       "KB-04-06-PRX-04-1               7.0               6.0               0.0   \n",
       "AG-00-45-PRX-00-1               0.0              11.0               0.0   \n",
       "KB-07-13-PRX-07-1              87.0             144.0               0.0   \n",
       "AM-04-08-PRX-04-1               2.0              10.0               0.0   \n",
       "AM-07-84-PRX-07-1              25.0              36.0               0.0   \n",
       "\n",
       "Gene ID              PF3D7_0523000                   \\\n",
       "Gene                          mdr1                    \n",
       "Mutation Name       mdr1-Ala821Thr  mdr1-Asn1042Asp   \n",
       "ExonicFunc        missense_variant missense_variant   \n",
       "AA Change                Ala821Thr       Asn1042Asp   \n",
       "Targeted                        No              Yes   \n",
       "KB-04-06-PRX-04-1              0.0              0.0   \n",
       "AG-00-45-PRX-00-1              0.0              0.0   \n",
       "KB-07-13-PRX-07-1              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0   \n",
       "\n",
       "Gene ID                                                             \\\n",
       "Gene                                                                 \n",
       "Mutation Name     mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "ExonicFunc          missense_variant&conservative_inframe_deletion   \n",
       "AA Change              Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "Targeted                                                        No   \n",
       "KB-04-06-PRX-04-1                                              0.0   \n",
       "AG-00-45-PRX-00-1                                              0.0   \n",
       "KB-07-13-PRX-07-1                                              0.0   \n",
       "AM-04-08-PRX-04-1                                              0.0   \n",
       "AM-07-84-PRX-07-1                                              0.0   \n",
       "\n",
       "Gene ID                                                           ...  \\\n",
       "Gene                                                              ...   \n",
       "Mutation Name                     mdr1-Asn647_Asn654delinsAspAsp  ...   \n",
       "ExonicFunc        missense_variant&conservative_inframe_deletion  ...   \n",
       "AA Change                              Asn647_Asn654delinsAspAsp  ...   \n",
       "Targeted                                                      No  ...   \n",
       "KB-04-06-PRX-04-1                                            0.0  ...   \n",
       "AG-00-45-PRX-00-1                                            0.0  ...   \n",
       "KB-07-13-PRX-07-1                                            0.0  ...   \n",
       "AM-04-08-PRX-04-1                                            0.0  ...   \n",
       "AM-07-84-PRX-07-1                                            0.0  ...   \n",
       "\n",
       "Gene ID              PF3D7_1408000                                    \\\n",
       "Gene                           PM2                                     \n",
       "Mutation Name        PM2-Asp440Asn    PM2-Gln442His    PM2-Ile401Leu   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Asp440Asn        Gln442His        Ile401Leu   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              0.0              4.0              0.0   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              0.0   \n",
       "KB-07-13-PRX-07-1              0.0            140.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0             21.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0             44.0              0.0   \n",
       "\n",
       "Gene ID                                                               \\\n",
       "Gene                                                                   \n",
       "Mutation Name        PM2-Leu277Val    PM2-Leu321Val    PM2-Leu398Val   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Leu277Val        Leu321Val        Leu398Val   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              0.0              0.0              0.0   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              0.0   \n",
       "KB-07-13-PRX-07-1              0.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                                                PF3D7_1408100  \\\n",
       "Gene                                                             PM3   \n",
       "Mutation Name        PM2-Phe350Val    PM2-Thr289Ile     PM3-Arg31Lys   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Phe350Val        Thr289Ile         Arg31Lys   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              0.0              0.0              0.0   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              0.0   \n",
       "KB-07-13-PRX-07-1            243.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                             \n",
       "Gene                                \n",
       "Mutation Name        PM3-Thr375Met  \n",
       "ExonicFunc        missense_variant  \n",
       "AA Change                Thr375Met  \n",
       "Targeted                        No  \n",
       "KB-04-06-PRX-04-1              0.0  \n",
       "AG-00-45-PRX-00-1              0.0  \n",
       "KB-07-13-PRX-07-1              0.0  \n",
       "AM-04-08-PRX-04-1              0.0  \n",
       "AM-07-84-PRX-07-1              0.0  \n",
       "\n",
       "[5 rows x 133 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RUN\n",
    "import pandas as pd\n",
    "mutation_counts = pd.read_csv(mutation_count_file,\n",
    "                              header=list(range(6)),\n",
    "                              index_col=0)\n",
    "mutation_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b2d630d6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Gene ID</th>\n",
       "      <th colspan=\"6\" halign=\"left\">PF3D7_0417200</th>\n",
       "      <th colspan=\"4\" halign=\"left\">PF3D7_0523000</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PF3D7_1408000</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PF3D7_1408100</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th colspan=\"6\" halign=\"left\">dhfr-ts</th>\n",
       "      <th colspan=\"4\" halign=\"left\">mdr1</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PM2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PM3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation Name</th>\n",
       "      <th>dhfr-ts-Ala16Val</th>\n",
       "      <th>dhfr-ts-Asn51Ile</th>\n",
       "      <th>dhfr-ts-Cys59Arg</th>\n",
       "      <th>dhfr-ts-Ile164Leu</th>\n",
       "      <th>dhfr-ts-Ser108Asn</th>\n",
       "      <th>dhfr-ts-Ser108Thr</th>\n",
       "      <th>mdr1-Ala821Thr</th>\n",
       "      <th>mdr1-Asn1042Asp</th>\n",
       "      <th>mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>mdr1-Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>PM2-Asp440Asn</th>\n",
       "      <th>PM2-Gln442His</th>\n",
       "      <th>PM2-Ile401Leu</th>\n",
       "      <th>PM2-Leu277Val</th>\n",
       "      <th>PM2-Leu321Val</th>\n",
       "      <th>PM2-Leu398Val</th>\n",
       "      <th>PM2-Phe350Val</th>\n",
       "      <th>PM2-Thr289Ile</th>\n",
       "      <th>PM3-Arg31Lys</th>\n",
       "      <th>PM3-Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ExonicFunc</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>...</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA Change</th>\n",
       "      <th>Ala16Val</th>\n",
       "      <th>Asn51Ile</th>\n",
       "      <th>Cys59Arg</th>\n",
       "      <th>Ile164Leu</th>\n",
       "      <th>Ser108Asn</th>\n",
       "      <th>Ser108Thr</th>\n",
       "      <th>Ala821Thr</th>\n",
       "      <th>Asn1042Asp</th>\n",
       "      <th>Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>Asp440Asn</th>\n",
       "      <th>Gln442His</th>\n",
       "      <th>Ile401Leu</th>\n",
       "      <th>Leu277Val</th>\n",
       "      <th>Leu321Val</th>\n",
       "      <th>Leu398Val</th>\n",
       "      <th>Phe350Val</th>\n",
       "      <th>Thr289Ile</th>\n",
       "      <th>Arg31Lys</th>\n",
       "      <th>Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Targeted</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>...</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KB-04-06-PRX-04-1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-00-45-PRX-00-1</th>\n",
       "      <td>24.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KB-07-13-PRX-07-1</th>\n",
       "      <td>179.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>131.0</td>\n",
       "      <td>759.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>...</td>\n",
       "      <td>140.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>243.0</td>\n",
       "      <td>245.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>183.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-04-08-PRX-04-1</th>\n",
       "      <td>17.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-07-84-PRX-07-1</th>\n",
       "      <td>34.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 133 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene ID              PF3D7_0417200                                    \\\n",
       "Gene                       dhfr-ts                                     \n",
       "Mutation Name     dhfr-ts-Ala16Val dhfr-ts-Asn51Ile dhfr-ts-Cys59Arg   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                 Ala16Val         Asn51Ile         Cys59Arg   \n",
       "Targeted                       Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1              2.0              2.0              2.0   \n",
       "AG-00-45-PRX-00-1             24.0             30.0             30.0   \n",
       "KB-07-13-PRX-07-1            179.0            254.0            254.0   \n",
       "AM-04-08-PRX-04-1             17.0             18.0             18.0   \n",
       "AM-07-84-PRX-07-1             34.0             51.0             51.0   \n",
       "\n",
       "Gene ID                                                                  \\\n",
       "Gene                                                                      \n",
       "Mutation Name     dhfr-ts-Ile164Leu dhfr-ts-Ser108Asn dhfr-ts-Ser108Thr   \n",
       "ExonicFunc         missense_variant  missense_variant  missense_variant   \n",
       "AA Change                 Ile164Leu         Ser108Asn         Ser108Thr   \n",
       "Targeted                        Yes               Yes               Yes   \n",
       "KB-04-06-PRX-04-1               7.0               6.0               6.0   \n",
       "AG-00-45-PRX-00-1               6.0              11.0              11.0   \n",
       "KB-07-13-PRX-07-1              87.0             144.0             144.0   \n",
       "AM-04-08-PRX-04-1              11.0              10.0              10.0   \n",
       "AM-07-84-PRX-07-1              25.0              36.0              36.0   \n",
       "\n",
       "Gene ID              PF3D7_0523000                   \\\n",
       "Gene                          mdr1                    \n",
       "Mutation Name       mdr1-Ala821Thr  mdr1-Asn1042Asp   \n",
       "ExonicFunc        missense_variant missense_variant   \n",
       "AA Change                Ala821Thr       Asn1042Asp   \n",
       "Targeted                        No              Yes   \n",
       "KB-04-06-PRX-04-1              4.0              9.0   \n",
       "AG-00-45-PRX-00-1              4.0            103.0   \n",
       "KB-07-13-PRX-07-1            131.0            759.0   \n",
       "AM-04-08-PRX-04-1             15.0             45.0   \n",
       "AM-07-84-PRX-07-1             26.0            106.0   \n",
       "\n",
       "Gene ID                                                             \\\n",
       "Gene                                                                 \n",
       "Mutation Name     mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "ExonicFunc          missense_variant&conservative_inframe_deletion   \n",
       "AA Change              Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "Targeted                                                        No   \n",
       "KB-04-06-PRX-04-1                                              1.0   \n",
       "AG-00-45-PRX-00-1                                             11.0   \n",
       "KB-07-13-PRX-07-1                                            221.0   \n",
       "AM-04-08-PRX-04-1                                              3.0   \n",
       "AM-07-84-PRX-07-1                                             37.0   \n",
       "\n",
       "Gene ID                                                           ...  \\\n",
       "Gene                                                              ...   \n",
       "Mutation Name                     mdr1-Asn647_Asn654delinsAspAsp  ...   \n",
       "ExonicFunc        missense_variant&conservative_inframe_deletion  ...   \n",
       "AA Change                              Asn647_Asn654delinsAspAsp  ...   \n",
       "Targeted                                                      No  ...   \n",
       "KB-04-06-PRX-04-1                                            1.0  ...   \n",
       "AG-00-45-PRX-00-1                                           11.0  ...   \n",
       "KB-07-13-PRX-07-1                                          221.0  ...   \n",
       "AM-04-08-PRX-04-1                                            3.0  ...   \n",
       "AM-07-84-PRX-07-1                                           37.0  ...   \n",
       "\n",
       "Gene ID              PF3D7_1408000                                    \\\n",
       "Gene                           PM2                                     \n",
       "Mutation Name        PM2-Asp440Asn    PM2-Gln442His    PM2-Ile401Leu   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Asp440Asn        Gln442His        Ile401Leu   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              4.0              4.0              3.0   \n",
       "AG-00-45-PRX-00-1              7.0              7.0              5.0   \n",
       "KB-07-13-PRX-07-1            140.0            140.0            103.0   \n",
       "AM-04-08-PRX-04-1             21.0             21.0             14.0   \n",
       "AM-07-84-PRX-07-1             44.0             44.0             29.0   \n",
       "\n",
       "Gene ID                                                               \\\n",
       "Gene                                                                   \n",
       "Mutation Name        PM2-Leu277Val    PM2-Leu321Val    PM2-Leu398Val   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Leu277Val        Leu321Val        Leu398Val   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              4.0              8.0              3.0   \n",
       "AG-00-45-PRX-00-1             22.0             15.0              5.0   \n",
       "KB-07-13-PRX-07-1            198.0             47.0            103.0   \n",
       "AM-04-08-PRX-04-1              9.0             23.0             14.0   \n",
       "AM-07-84-PRX-07-1             27.0             32.0             29.0   \n",
       "\n",
       "Gene ID                                                PF3D7_1408100  \\\n",
       "Gene                                                             PM3   \n",
       "Mutation Name        PM2-Phe350Val    PM2-Thr289Ile     PM3-Arg31Lys   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Phe350Val        Thr289Ile         Arg31Lys   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              1.0             12.0              2.0   \n",
       "AG-00-45-PRX-00-1             21.0             37.0              7.0   \n",
       "KB-07-13-PRX-07-1            243.0            245.0            188.0   \n",
       "AM-04-08-PRX-04-1              6.0             32.0             12.0   \n",
       "AM-07-84-PRX-07-1             28.0             59.0             38.0   \n",
       "\n",
       "Gene ID                             \n",
       "Gene                                \n",
       "Mutation Name        PM3-Thr375Met  \n",
       "ExonicFunc        missense_variant  \n",
       "AA Change                Thr375Met  \n",
       "Targeted                        No  \n",
       "KB-04-06-PRX-04-1              0.0  \n",
       "AG-00-45-PRX-00-1             11.0  \n",
       "KB-07-13-PRX-07-1            183.0  \n",
       "AM-04-08-PRX-04-1              7.0  \n",
       "AM-07-84-PRX-07-1             31.0  \n",
       "\n",
       "[5 rows x 133 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RUN\n",
    "mutation_coverage = pd.read_csv(mutation_coverage_file,\n",
    "                                index_col=0,\n",
    "                                header=list(range(6)))\n",
    "mutation_coverage.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f3d6741",
   "metadata": {},
   "source": [
    "### Set your filters   \n",
    "1.  **min_coverage**: how many UMIs are needed to for a genomic position for a sample to reliably call genotypes. If we set min_coverage = 10, any locus within a sample that is covered below this threshold will have an NA genotype.\n",
    "2.  **min_count**: if a genomic position have enough coverage, how many UMIs supporting an ALT allele call is needed for a reliable call. If we set min_count = 2, any mutation with an  call that has less than 2 UMIs supporting the ALT call will revert to REF.\n",
    "3.  **min_freq**: a minimum within sample allele frequency threshold to consider a variant valid. If set to 0.01, for example, a variant locus in a sample that is at 0.005 frequency for the ALT allele within the sample, the locus would be called REF, if the within sample AF is between 0.01 and 0.99, it would be considered HET, and if > 0.99, it would be homozygous ALT."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a0f680c",
   "metadata": {},
   "source": [
    "#### Example cell\n",
    "```python\n",
    "# filter mutation counts for minimum count parameter\n",
    "# by setting counts to zero if it is below threshold\n",
    "min_count = 2\n",
    "# filter loci without enough coverage by setting\n",
    "# coverage to zero if it is below threshold\n",
    "min_coverage = 10\n",
    "# call genotypes using the minimum within sample\n",
    "# allele frequency\n",
    "min_freq = 0\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b05fb266",
   "metadata": {},
   "outputs": [],
   "source": [
    "# USER INPUT \n",
    "\n",
    "# filter mutation counts for minimum count parameter\n",
    "# by setting counts to zero if it is below threshold\n",
    "min_count = 2\n",
    "# filter loci without enough coverage by setting\n",
    "# coverage to zero if it is below threshold\n",
    "min_coverage = 10\n",
    "# call genotypes using the minimum within sample\n",
    "# allele frequency\n",
    "min_freq = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1c87661e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['genotypes', 'prevalences', 'wsaf', 'filtered_mutation_counts', 'filtered_mutation_coverage'])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RUN\n",
    "\n",
    "# import the PCA module which has genotype calling and\n",
    "# filtering functions \n",
    "import PCA\n",
    "\n",
    "gt_calls = PCA.call_genotypes(mutation_counts, mutation_coverage,\n",
    "                              min_count, min_coverage, min_freq)\n",
    "gt_calls.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94fbe8fd",
   "metadata": {},
   "source": [
    "### What are the dataframes generated by call_genotypes function and how  are they generated?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab506a25",
   "metadata": {},
   "source": [
    "**filtered_mutation_counts**: take the mutation_counts table, if a cell's value is below *min_count*, reset that cell's value to zero, otherwise leave as is.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b2649811",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Gene ID</th>\n",
       "      <th colspan=\"6\" halign=\"left\">PF3D7_0417200</th>\n",
       "      <th colspan=\"4\" halign=\"left\">PF3D7_0523000</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PF3D7_1408000</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PF3D7_1408100</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th colspan=\"6\" halign=\"left\">dhfr-ts</th>\n",
       "      <th colspan=\"4\" halign=\"left\">mdr1</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PM2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PM3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation Name</th>\n",
       "      <th>dhfr-ts-Ala16Val</th>\n",
       "      <th>dhfr-ts-Asn51Ile</th>\n",
       "      <th>dhfr-ts-Cys59Arg</th>\n",
       "      <th>dhfr-ts-Ile164Leu</th>\n",
       "      <th>dhfr-ts-Ser108Asn</th>\n",
       "      <th>dhfr-ts-Ser108Thr</th>\n",
       "      <th>mdr1-Ala821Thr</th>\n",
       "      <th>mdr1-Asn1042Asp</th>\n",
       "      <th>mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>mdr1-Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>PM2-Asp440Asn</th>\n",
       "      <th>PM2-Gln442His</th>\n",
       "      <th>PM2-Ile401Leu</th>\n",
       "      <th>PM2-Leu277Val</th>\n",
       "      <th>PM2-Leu321Val</th>\n",
       "      <th>PM2-Leu398Val</th>\n",
       "      <th>PM2-Phe350Val</th>\n",
       "      <th>PM2-Thr289Ile</th>\n",
       "      <th>PM3-Arg31Lys</th>\n",
       "      <th>PM3-Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ExonicFunc</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>...</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA Change</th>\n",
       "      <th>Ala16Val</th>\n",
       "      <th>Asn51Ile</th>\n",
       "      <th>Cys59Arg</th>\n",
       "      <th>Ile164Leu</th>\n",
       "      <th>Ser108Asn</th>\n",
       "      <th>Ser108Thr</th>\n",
       "      <th>Ala821Thr</th>\n",
       "      <th>Asn1042Asp</th>\n",
       "      <th>Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>Asp440Asn</th>\n",
       "      <th>Gln442His</th>\n",
       "      <th>Ile401Leu</th>\n",
       "      <th>Leu277Val</th>\n",
       "      <th>Leu321Val</th>\n",
       "      <th>Leu398Val</th>\n",
       "      <th>Phe350Val</th>\n",
       "      <th>Thr289Ile</th>\n",
       "      <th>Arg31Lys</th>\n",
       "      <th>Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Targeted</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>...</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KB-04-06-PRX-04-1</th>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-00-45-PRX-00-1</th>\n",
       "      <td>0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KB-07-13-PRX-07-1</th>\n",
       "      <td>0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>243.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-04-08-PRX-04-1</th>\n",
       "      <td>0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-07-84-PRX-07-1</th>\n",
       "      <td>0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 133 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene ID              PF3D7_0417200                                    \\\n",
       "Gene                       dhfr-ts                                     \n",
       "Mutation Name     dhfr-ts-Ala16Val dhfr-ts-Asn51Ile dhfr-ts-Cys59Arg   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                 Ala16Val         Asn51Ile         Cys59Arg   \n",
       "Targeted                       Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1                0              2.0              2.0   \n",
       "AG-00-45-PRX-00-1                0             30.0             30.0   \n",
       "KB-07-13-PRX-07-1                0            254.0            254.0   \n",
       "AM-04-08-PRX-04-1                0             18.0             18.0   \n",
       "AM-07-84-PRX-07-1                0             51.0             51.0   \n",
       "\n",
       "Gene ID                                                                  \\\n",
       "Gene                                                                      \n",
       "Mutation Name     dhfr-ts-Ile164Leu dhfr-ts-Ser108Asn dhfr-ts-Ser108Thr   \n",
       "ExonicFunc         missense_variant  missense_variant  missense_variant   \n",
       "AA Change                 Ile164Leu         Ser108Asn         Ser108Thr   \n",
       "Targeted                        Yes               Yes               Yes   \n",
       "KB-04-06-PRX-04-1               7.0               6.0                 0   \n",
       "AG-00-45-PRX-00-1               0.0              11.0                 0   \n",
       "KB-07-13-PRX-07-1              87.0             144.0                 0   \n",
       "AM-04-08-PRX-04-1               2.0              10.0                 0   \n",
       "AM-07-84-PRX-07-1              25.0              36.0                 0   \n",
       "\n",
       "Gene ID              PF3D7_0523000                   \\\n",
       "Gene                          mdr1                    \n",
       "Mutation Name       mdr1-Ala821Thr  mdr1-Asn1042Asp   \n",
       "ExonicFunc        missense_variant missense_variant   \n",
       "AA Change                Ala821Thr       Asn1042Asp   \n",
       "Targeted                        No              Yes   \n",
       "KB-04-06-PRX-04-1              0.0                0   \n",
       "AG-00-45-PRX-00-1              0.0                0   \n",
       "KB-07-13-PRX-07-1              0.0                0   \n",
       "AM-04-08-PRX-04-1              0.0                0   \n",
       "AM-07-84-PRX-07-1              0.0                0   \n",
       "\n",
       "Gene ID                                                             \\\n",
       "Gene                                                                 \n",
       "Mutation Name     mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "ExonicFunc          missense_variant&conservative_inframe_deletion   \n",
       "AA Change              Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "Targeted                                                        No   \n",
       "KB-04-06-PRX-04-1                                              0.0   \n",
       "AG-00-45-PRX-00-1                                              0.0   \n",
       "KB-07-13-PRX-07-1                                              0.0   \n",
       "AM-04-08-PRX-04-1                                              0.0   \n",
       "AM-07-84-PRX-07-1                                              0.0   \n",
       "\n",
       "Gene ID                                                           ...  \\\n",
       "Gene                                                              ...   \n",
       "Mutation Name                     mdr1-Asn647_Asn654delinsAspAsp  ...   \n",
       "ExonicFunc        missense_variant&conservative_inframe_deletion  ...   \n",
       "AA Change                              Asn647_Asn654delinsAspAsp  ...   \n",
       "Targeted                                                      No  ...   \n",
       "KB-04-06-PRX-04-1                                            0.0  ...   \n",
       "AG-00-45-PRX-00-1                                            0.0  ...   \n",
       "KB-07-13-PRX-07-1                                            0.0  ...   \n",
       "AM-04-08-PRX-04-1                                            0.0  ...   \n",
       "AM-07-84-PRX-07-1                                            0.0  ...   \n",
       "\n",
       "Gene ID              PF3D7_1408000                                    \\\n",
       "Gene                           PM2                                     \n",
       "Mutation Name        PM2-Asp440Asn    PM2-Gln442His    PM2-Ile401Leu   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Asp440Asn        Gln442His        Ile401Leu   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              0.0              4.0              0.0   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              0.0   \n",
       "KB-07-13-PRX-07-1              0.0            140.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0             21.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0             44.0              0.0   \n",
       "\n",
       "Gene ID                                                               \\\n",
       "Gene                                                                   \n",
       "Mutation Name        PM2-Leu277Val    PM2-Leu321Val    PM2-Leu398Val   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Leu277Val        Leu321Val        Leu398Val   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              0.0              0.0              0.0   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              0.0   \n",
       "KB-07-13-PRX-07-1              0.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                                                PF3D7_1408100  \\\n",
       "Gene                                                             PM3   \n",
       "Mutation Name        PM2-Phe350Val    PM2-Thr289Ile     PM3-Arg31Lys   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Phe350Val        Thr289Ile         Arg31Lys   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              0.0              0.0              0.0   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              0.0   \n",
       "KB-07-13-PRX-07-1            243.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                             \n",
       "Gene                                \n",
       "Mutation Name        PM3-Thr375Met  \n",
       "ExonicFunc        missense_variant  \n",
       "AA Change                Thr375Met  \n",
       "Targeted                        No  \n",
       "KB-04-06-PRX-04-1              0.0  \n",
       "AG-00-45-PRX-00-1              0.0  \n",
       "KB-07-13-PRX-07-1              0.0  \n",
       "AM-04-08-PRX-04-1              0.0  \n",
       "AM-07-84-PRX-07-1              0.0  \n",
       "\n",
       "[5 rows x 133 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RUN\n",
    "filtered_mutation_counts = gt_calls[\"filtered_mutation_counts\"]\n",
    "filtered_mutation_counts.to_csv(os.path.join(\n",
    "        wdir, \"filtered_alternate_AA_table.csv\"))\n",
    "filtered_mutation_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35de8bb0",
   "metadata": {},
   "source": [
    "**filtered_mutation_coverage**: take the mutation_coverage table, if a cell's value is below *min_coverage*, reset that cell's value to zero, otherwise leave as is."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6e1b3544",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Gene ID</th>\n",
       "      <th colspan=\"6\" halign=\"left\">PF3D7_0417200</th>\n",
       "      <th colspan=\"4\" halign=\"left\">PF3D7_0523000</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PF3D7_1408000</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PF3D7_1408100</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th colspan=\"6\" halign=\"left\">dhfr-ts</th>\n",
       "      <th colspan=\"4\" halign=\"left\">mdr1</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PM2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PM3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation Name</th>\n",
       "      <th>dhfr-ts-Ala16Val</th>\n",
       "      <th>dhfr-ts-Asn51Ile</th>\n",
       "      <th>dhfr-ts-Cys59Arg</th>\n",
       "      <th>dhfr-ts-Ile164Leu</th>\n",
       "      <th>dhfr-ts-Ser108Asn</th>\n",
       "      <th>dhfr-ts-Ser108Thr</th>\n",
       "      <th>mdr1-Ala821Thr</th>\n",
       "      <th>mdr1-Asn1042Asp</th>\n",
       "      <th>mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>mdr1-Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>PM2-Asp440Asn</th>\n",
       "      <th>PM2-Gln442His</th>\n",
       "      <th>PM2-Ile401Leu</th>\n",
       "      <th>PM2-Leu277Val</th>\n",
       "      <th>PM2-Leu321Val</th>\n",
       "      <th>PM2-Leu398Val</th>\n",
       "      <th>PM2-Phe350Val</th>\n",
       "      <th>PM2-Thr289Ile</th>\n",
       "      <th>PM3-Arg31Lys</th>\n",
       "      <th>PM3-Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ExonicFunc</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>...</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA Change</th>\n",
       "      <th>Ala16Val</th>\n",
       "      <th>Asn51Ile</th>\n",
       "      <th>Cys59Arg</th>\n",
       "      <th>Ile164Leu</th>\n",
       "      <th>Ser108Asn</th>\n",
       "      <th>Ser108Thr</th>\n",
       "      <th>Ala821Thr</th>\n",
       "      <th>Asn1042Asp</th>\n",
       "      <th>Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>Asp440Asn</th>\n",
       "      <th>Gln442His</th>\n",
       "      <th>Ile401Leu</th>\n",
       "      <th>Leu277Val</th>\n",
       "      <th>Leu321Val</th>\n",
       "      <th>Leu398Val</th>\n",
       "      <th>Phe350Val</th>\n",
       "      <th>Thr289Ile</th>\n",
       "      <th>Arg31Lys</th>\n",
       "      <th>Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Targeted</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>...</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KB-04-06-PRX-04-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-00-45-PRX-00-1</th>\n",
       "      <td>24.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KB-07-13-PRX-07-1</th>\n",
       "      <td>179.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>131.0</td>\n",
       "      <td>759.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>...</td>\n",
       "      <td>140.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>243.0</td>\n",
       "      <td>245.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>183.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-04-08-PRX-04-1</th>\n",
       "      <td>17.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-07-84-PRX-07-1</th>\n",
       "      <td>34.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 133 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene ID              PF3D7_0417200                                    \\\n",
       "Gene                       dhfr-ts                                     \n",
       "Mutation Name     dhfr-ts-Ala16Val dhfr-ts-Asn51Ile dhfr-ts-Cys59Arg   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                 Ala16Val         Asn51Ile         Cys59Arg   \n",
       "Targeted                       Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1              0.0              0.0              0.0   \n",
       "AG-00-45-PRX-00-1             24.0             30.0             30.0   \n",
       "KB-07-13-PRX-07-1            179.0            254.0            254.0   \n",
       "AM-04-08-PRX-04-1             17.0             18.0             18.0   \n",
       "AM-07-84-PRX-07-1             34.0             51.0             51.0   \n",
       "\n",
       "Gene ID                                                                  \\\n",
       "Gene                                                                      \n",
       "Mutation Name     dhfr-ts-Ile164Leu dhfr-ts-Ser108Asn dhfr-ts-Ser108Thr   \n",
       "ExonicFunc         missense_variant  missense_variant  missense_variant   \n",
       "AA Change                 Ile164Leu         Ser108Asn         Ser108Thr   \n",
       "Targeted                        Yes               Yes               Yes   \n",
       "KB-04-06-PRX-04-1               0.0               0.0               0.0   \n",
       "AG-00-45-PRX-00-1               0.0              11.0              11.0   \n",
       "KB-07-13-PRX-07-1              87.0             144.0             144.0   \n",
       "AM-04-08-PRX-04-1              11.0              10.0              10.0   \n",
       "AM-07-84-PRX-07-1              25.0              36.0              36.0   \n",
       "\n",
       "Gene ID              PF3D7_0523000                   \\\n",
       "Gene                          mdr1                    \n",
       "Mutation Name       mdr1-Ala821Thr  mdr1-Asn1042Asp   \n",
       "ExonicFunc        missense_variant missense_variant   \n",
       "AA Change                Ala821Thr       Asn1042Asp   \n",
       "Targeted                        No              Yes   \n",
       "KB-04-06-PRX-04-1              0.0              0.0   \n",
       "AG-00-45-PRX-00-1              0.0            103.0   \n",
       "KB-07-13-PRX-07-1            131.0            759.0   \n",
       "AM-04-08-PRX-04-1             15.0             45.0   \n",
       "AM-07-84-PRX-07-1             26.0            106.0   \n",
       "\n",
       "Gene ID                                                             \\\n",
       "Gene                                                                 \n",
       "Mutation Name     mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "ExonicFunc          missense_variant&conservative_inframe_deletion   \n",
       "AA Change              Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "Targeted                                                        No   \n",
       "KB-04-06-PRX-04-1                                              0.0   \n",
       "AG-00-45-PRX-00-1                                             11.0   \n",
       "KB-07-13-PRX-07-1                                            221.0   \n",
       "AM-04-08-PRX-04-1                                              0.0   \n",
       "AM-07-84-PRX-07-1                                             37.0   \n",
       "\n",
       "Gene ID                                                           ...  \\\n",
       "Gene                                                              ...   \n",
       "Mutation Name                     mdr1-Asn647_Asn654delinsAspAsp  ...   \n",
       "ExonicFunc        missense_variant&conservative_inframe_deletion  ...   \n",
       "AA Change                              Asn647_Asn654delinsAspAsp  ...   \n",
       "Targeted                                                      No  ...   \n",
       "KB-04-06-PRX-04-1                                            0.0  ...   \n",
       "AG-00-45-PRX-00-1                                           11.0  ...   \n",
       "KB-07-13-PRX-07-1                                          221.0  ...   \n",
       "AM-04-08-PRX-04-1                                            0.0  ...   \n",
       "AM-07-84-PRX-07-1                                           37.0  ...   \n",
       "\n",
       "Gene ID              PF3D7_1408000                                    \\\n",
       "Gene                           PM2                                     \n",
       "Mutation Name        PM2-Asp440Asn    PM2-Gln442His    PM2-Ile401Leu   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Asp440Asn        Gln442His        Ile401Leu   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              0.0              0.0              0.0   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              0.0   \n",
       "KB-07-13-PRX-07-1            140.0            140.0            103.0   \n",
       "AM-04-08-PRX-04-1             21.0             21.0             14.0   \n",
       "AM-07-84-PRX-07-1             44.0             44.0             29.0   \n",
       "\n",
       "Gene ID                                                               \\\n",
       "Gene                                                                   \n",
       "Mutation Name        PM2-Leu277Val    PM2-Leu321Val    PM2-Leu398Val   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Leu277Val        Leu321Val        Leu398Val   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              0.0              0.0              0.0   \n",
       "AG-00-45-PRX-00-1             22.0             15.0              0.0   \n",
       "KB-07-13-PRX-07-1            198.0             47.0            103.0   \n",
       "AM-04-08-PRX-04-1              0.0             23.0             14.0   \n",
       "AM-07-84-PRX-07-1             27.0             32.0             29.0   \n",
       "\n",
       "Gene ID                                                PF3D7_1408100  \\\n",
       "Gene                                                             PM3   \n",
       "Mutation Name        PM2-Phe350Val    PM2-Thr289Ile     PM3-Arg31Lys   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Phe350Val        Thr289Ile         Arg31Lys   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              0.0             12.0              0.0   \n",
       "AG-00-45-PRX-00-1             21.0             37.0              0.0   \n",
       "KB-07-13-PRX-07-1            243.0            245.0            188.0   \n",
       "AM-04-08-PRX-04-1              0.0             32.0             12.0   \n",
       "AM-07-84-PRX-07-1             28.0             59.0             38.0   \n",
       "\n",
       "Gene ID                             \n",
       "Gene                                \n",
       "Mutation Name        PM3-Thr375Met  \n",
       "ExonicFunc        missense_variant  \n",
       "AA Change                Thr375Met  \n",
       "Targeted                        No  \n",
       "KB-04-06-PRX-04-1              0.0  \n",
       "AG-00-45-PRX-00-1             11.0  \n",
       "KB-07-13-PRX-07-1            183.0  \n",
       "AM-04-08-PRX-04-1              0.0  \n",
       "AM-07-84-PRX-07-1             31.0  \n",
       "\n",
       "[5 rows x 133 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RUN\n",
    "filtered_mutation_coverage = gt_calls[\"filtered_mutation_coverage\"]\n",
    "filtered_mutation_coverage.to_csv(os.path.join(\n",
    "        wdir, \"filtered_coverage_AA_table.csv\"))\n",
    "filtered_mutation_coverage.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccaafa91",
   "metadata": {},
   "source": [
    "**wsaf**: divide *filtered_mutation_counts* table by *filtered_mutation_coverage* table, yielding within sample allele frequencies.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fcc97d5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Gene ID</th>\n",
       "      <th colspan=\"6\" halign=\"left\">PF3D7_0417200</th>\n",
       "      <th colspan=\"4\" halign=\"left\">PF3D7_0523000</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PF3D7_1408000</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PF3D7_1408100</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th colspan=\"6\" halign=\"left\">dhfr-ts</th>\n",
       "      <th colspan=\"4\" halign=\"left\">mdr1</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PM2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PM3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation Name</th>\n",
       "      <th>dhfr-ts-Ala16Val</th>\n",
       "      <th>dhfr-ts-Asn51Ile</th>\n",
       "      <th>dhfr-ts-Cys59Arg</th>\n",
       "      <th>dhfr-ts-Ile164Leu</th>\n",
       "      <th>dhfr-ts-Ser108Asn</th>\n",
       "      <th>dhfr-ts-Ser108Thr</th>\n",
       "      <th>mdr1-Ala821Thr</th>\n",
       "      <th>mdr1-Asn1042Asp</th>\n",
       "      <th>mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>mdr1-Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>PM2-Asp440Asn</th>\n",
       "      <th>PM2-Gln442His</th>\n",
       "      <th>PM2-Ile401Leu</th>\n",
       "      <th>PM2-Leu277Val</th>\n",
       "      <th>PM2-Leu321Val</th>\n",
       "      <th>PM2-Leu398Val</th>\n",
       "      <th>PM2-Phe350Val</th>\n",
       "      <th>PM2-Thr289Ile</th>\n",
       "      <th>PM3-Arg31Lys</th>\n",
       "      <th>PM3-Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ExonicFunc</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>...</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA Change</th>\n",
       "      <th>Ala16Val</th>\n",
       "      <th>Asn51Ile</th>\n",
       "      <th>Cys59Arg</th>\n",
       "      <th>Ile164Leu</th>\n",
       "      <th>Ser108Asn</th>\n",
       "      <th>Ser108Thr</th>\n",
       "      <th>Ala821Thr</th>\n",
       "      <th>Asn1042Asp</th>\n",
       "      <th>Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>Asp440Asn</th>\n",
       "      <th>Gln442His</th>\n",
       "      <th>Ile401Leu</th>\n",
       "      <th>Leu277Val</th>\n",
       "      <th>Leu321Val</th>\n",
       "      <th>Leu398Val</th>\n",
       "      <th>Phe350Val</th>\n",
       "      <th>Thr289Ile</th>\n",
       "      <th>Arg31Lys</th>\n",
       "      <th>Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Targeted</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>...</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KB-04-06-PRX-04-1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-00-45-PRX-00-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KB-07-13-PRX-07-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-04-08-PRX-04-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-07-84-PRX-07-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 133 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene ID              PF3D7_0417200                                    \\\n",
       "Gene                       dhfr-ts                                     \n",
       "Mutation Name     dhfr-ts-Ala16Val dhfr-ts-Asn51Ile dhfr-ts-Cys59Arg   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                 Ala16Val         Asn51Ile         Cys59Arg   \n",
       "Targeted                       Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              1.0              1.0   \n",
       "KB-07-13-PRX-07-1              0.0              1.0              1.0   \n",
       "AM-04-08-PRX-04-1              0.0              1.0              1.0   \n",
       "AM-07-84-PRX-07-1              0.0              1.0              1.0   \n",
       "\n",
       "Gene ID                                                                  \\\n",
       "Gene                                                                      \n",
       "Mutation Name     dhfr-ts-Ile164Leu dhfr-ts-Ser108Asn dhfr-ts-Ser108Thr   \n",
       "ExonicFunc         missense_variant  missense_variant  missense_variant   \n",
       "AA Change                 Ile164Leu         Ser108Asn         Ser108Thr   \n",
       "Targeted                        Yes               Yes               Yes   \n",
       "KB-04-06-PRX-04-1               NaN               NaN               NaN   \n",
       "AG-00-45-PRX-00-1               NaN               1.0               0.0   \n",
       "KB-07-13-PRX-07-1          1.000000               1.0               0.0   \n",
       "AM-04-08-PRX-04-1          0.181818               1.0               0.0   \n",
       "AM-07-84-PRX-07-1          1.000000               1.0               0.0   \n",
       "\n",
       "Gene ID              PF3D7_0523000                   \\\n",
       "Gene                          mdr1                    \n",
       "Mutation Name       mdr1-Ala821Thr  mdr1-Asn1042Asp   \n",
       "ExonicFunc        missense_variant missense_variant   \n",
       "AA Change                Ala821Thr       Asn1042Asp   \n",
       "Targeted                        No              Yes   \n",
       "KB-04-06-PRX-04-1              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              NaN              0.0   \n",
       "KB-07-13-PRX-07-1              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0   \n",
       "\n",
       "Gene ID                                                             \\\n",
       "Gene                                                                 \n",
       "Mutation Name     mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "ExonicFunc          missense_variant&conservative_inframe_deletion   \n",
       "AA Change              Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "Targeted                                                        No   \n",
       "KB-04-06-PRX-04-1                                              NaN   \n",
       "AG-00-45-PRX-00-1                                              0.0   \n",
       "KB-07-13-PRX-07-1                                              0.0   \n",
       "AM-04-08-PRX-04-1                                              NaN   \n",
       "AM-07-84-PRX-07-1                                              0.0   \n",
       "\n",
       "Gene ID                                                           ...  \\\n",
       "Gene                                                              ...   \n",
       "Mutation Name                     mdr1-Asn647_Asn654delinsAspAsp  ...   \n",
       "ExonicFunc        missense_variant&conservative_inframe_deletion  ...   \n",
       "AA Change                              Asn647_Asn654delinsAspAsp  ...   \n",
       "Targeted                                                      No  ...   \n",
       "KB-04-06-PRX-04-1                                            NaN  ...   \n",
       "AG-00-45-PRX-00-1                                            0.0  ...   \n",
       "KB-07-13-PRX-07-1                                            0.0  ...   \n",
       "AM-04-08-PRX-04-1                                            NaN  ...   \n",
       "AM-07-84-PRX-07-1                                            0.0  ...   \n",
       "\n",
       "Gene ID              PF3D7_1408000                                    \\\n",
       "Gene                           PM2                                     \n",
       "Mutation Name        PM2-Asp440Asn    PM2-Gln442His    PM2-Ile401Leu   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Asp440Asn        Gln442His        Ile401Leu   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              NaN              NaN              NaN   \n",
       "KB-07-13-PRX-07-1              0.0              1.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              1.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              1.0              0.0   \n",
       "\n",
       "Gene ID                                                               \\\n",
       "Gene                                                                   \n",
       "Mutation Name        PM2-Leu277Val    PM2-Leu321Val    PM2-Leu398Val   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Leu277Val        Leu321Val        Leu398Val   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              NaN   \n",
       "KB-07-13-PRX-07-1              0.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              NaN              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                                                PF3D7_1408100  \\\n",
       "Gene                                                             PM3   \n",
       "Mutation Name        PM2-Phe350Val    PM2-Thr289Ile     PM3-Arg31Lys   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Phe350Val        Thr289Ile         Arg31Lys   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              NaN              0.0              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              NaN   \n",
       "KB-07-13-PRX-07-1              1.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              NaN              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                             \n",
       "Gene                                \n",
       "Mutation Name        PM3-Thr375Met  \n",
       "ExonicFunc        missense_variant  \n",
       "AA Change                Thr375Met  \n",
       "Targeted                        No  \n",
       "KB-04-06-PRX-04-1              NaN  \n",
       "AG-00-45-PRX-00-1              0.0  \n",
       "KB-07-13-PRX-07-1              0.0  \n",
       "AM-04-08-PRX-04-1              NaN  \n",
       "AM-07-84-PRX-07-1              0.0  \n",
       "\n",
       "[5 rows x 133 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RUN\n",
    "freq = gt_calls[\"wsaf\"]\n",
    "freq.to_csv(os.path.join(\n",
    "        wdir, \"within_sample_allele_frequencies.csv\"))\n",
    "freq.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e976119",
   "metadata": {},
   "source": [
    "**genotypes**: take the *wsaf* table, if a cell's value is less than *min_freq* set the genotype value to 0 (homozygous wild type); if the cell's value is more than (*1 - min_freq*) set the genotype value to 2 (homozygous mutant), if the cell's value is between *min_freq* and (*1 - min_freq*) set the genotype value to 1 (heterozygous/mixed).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3c4efc52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Gene ID</th>\n",
       "      <th colspan=\"6\" halign=\"left\">PF3D7_0417200</th>\n",
       "      <th colspan=\"4\" halign=\"left\">PF3D7_0523000</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PF3D7_1408000</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PF3D7_1408100</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th colspan=\"6\" halign=\"left\">dhfr-ts</th>\n",
       "      <th colspan=\"4\" halign=\"left\">mdr1</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PM2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PM3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation Name</th>\n",
       "      <th>dhfr-ts-Ala16Val</th>\n",
       "      <th>dhfr-ts-Asn51Ile</th>\n",
       "      <th>dhfr-ts-Cys59Arg</th>\n",
       "      <th>dhfr-ts-Ile164Leu</th>\n",
       "      <th>dhfr-ts-Ser108Asn</th>\n",
       "      <th>dhfr-ts-Ser108Thr</th>\n",
       "      <th>mdr1-Ala821Thr</th>\n",
       "      <th>mdr1-Asn1042Asp</th>\n",
       "      <th>mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>mdr1-Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>PM2-Asp440Asn</th>\n",
       "      <th>PM2-Gln442His</th>\n",
       "      <th>PM2-Ile401Leu</th>\n",
       "      <th>PM2-Leu277Val</th>\n",
       "      <th>PM2-Leu321Val</th>\n",
       "      <th>PM2-Leu398Val</th>\n",
       "      <th>PM2-Phe350Val</th>\n",
       "      <th>PM2-Thr289Ile</th>\n",
       "      <th>PM3-Arg31Lys</th>\n",
       "      <th>PM3-Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ExonicFunc</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>...</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA Change</th>\n",
       "      <th>Ala16Val</th>\n",
       "      <th>Asn51Ile</th>\n",
       "      <th>Cys59Arg</th>\n",
       "      <th>Ile164Leu</th>\n",
       "      <th>Ser108Asn</th>\n",
       "      <th>Ser108Thr</th>\n",
       "      <th>Ala821Thr</th>\n",
       "      <th>Asn1042Asp</th>\n",
       "      <th>Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>Asp440Asn</th>\n",
       "      <th>Gln442His</th>\n",
       "      <th>Ile401Leu</th>\n",
       "      <th>Leu277Val</th>\n",
       "      <th>Leu321Val</th>\n",
       "      <th>Leu398Val</th>\n",
       "      <th>Phe350Val</th>\n",
       "      <th>Thr289Ile</th>\n",
       "      <th>Arg31Lys</th>\n",
       "      <th>Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Targeted</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>...</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KB-04-06-PRX-04-1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-00-45-PRX-00-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KB-07-13-PRX-07-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-04-08-PRX-04-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-07-84-PRX-07-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 133 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene ID              PF3D7_0417200                                    \\\n",
       "Gene                       dhfr-ts                                     \n",
       "Mutation Name     dhfr-ts-Ala16Val dhfr-ts-Asn51Ile dhfr-ts-Cys59Arg   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                 Ala16Val         Asn51Ile         Cys59Arg   \n",
       "Targeted                       Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              2.0              2.0   \n",
       "KB-07-13-PRX-07-1              0.0              2.0              2.0   \n",
       "AM-04-08-PRX-04-1              0.0              2.0              2.0   \n",
       "AM-07-84-PRX-07-1              0.0              2.0              2.0   \n",
       "\n",
       "Gene ID                                                                  \\\n",
       "Gene                                                                      \n",
       "Mutation Name     dhfr-ts-Ile164Leu dhfr-ts-Ser108Asn dhfr-ts-Ser108Thr   \n",
       "ExonicFunc         missense_variant  missense_variant  missense_variant   \n",
       "AA Change                 Ile164Leu         Ser108Asn         Ser108Thr   \n",
       "Targeted                        Yes               Yes               Yes   \n",
       "KB-04-06-PRX-04-1               NaN               NaN               NaN   \n",
       "AG-00-45-PRX-00-1               NaN               2.0               0.0   \n",
       "KB-07-13-PRX-07-1               2.0               2.0               0.0   \n",
       "AM-04-08-PRX-04-1               1.0               2.0               0.0   \n",
       "AM-07-84-PRX-07-1               2.0               2.0               0.0   \n",
       "\n",
       "Gene ID              PF3D7_0523000                   \\\n",
       "Gene                          mdr1                    \n",
       "Mutation Name       mdr1-Ala821Thr  mdr1-Asn1042Asp   \n",
       "ExonicFunc        missense_variant missense_variant   \n",
       "AA Change                Ala821Thr       Asn1042Asp   \n",
       "Targeted                        No              Yes   \n",
       "KB-04-06-PRX-04-1              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              NaN              0.0   \n",
       "KB-07-13-PRX-07-1              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0   \n",
       "\n",
       "Gene ID                                                             \\\n",
       "Gene                                                                 \n",
       "Mutation Name     mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "ExonicFunc          missense_variant&conservative_inframe_deletion   \n",
       "AA Change              Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "Targeted                                                        No   \n",
       "KB-04-06-PRX-04-1                                              NaN   \n",
       "AG-00-45-PRX-00-1                                              0.0   \n",
       "KB-07-13-PRX-07-1                                              0.0   \n",
       "AM-04-08-PRX-04-1                                              NaN   \n",
       "AM-07-84-PRX-07-1                                              0.0   \n",
       "\n",
       "Gene ID                                                           ...  \\\n",
       "Gene                                                              ...   \n",
       "Mutation Name                     mdr1-Asn647_Asn654delinsAspAsp  ...   \n",
       "ExonicFunc        missense_variant&conservative_inframe_deletion  ...   \n",
       "AA Change                              Asn647_Asn654delinsAspAsp  ...   \n",
       "Targeted                                                      No  ...   \n",
       "KB-04-06-PRX-04-1                                            NaN  ...   \n",
       "AG-00-45-PRX-00-1                                            0.0  ...   \n",
       "KB-07-13-PRX-07-1                                            0.0  ...   \n",
       "AM-04-08-PRX-04-1                                            NaN  ...   \n",
       "AM-07-84-PRX-07-1                                            0.0  ...   \n",
       "\n",
       "Gene ID              PF3D7_1408000                                    \\\n",
       "Gene                           PM2                                     \n",
       "Mutation Name        PM2-Asp440Asn    PM2-Gln442His    PM2-Ile401Leu   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Asp440Asn        Gln442His        Ile401Leu   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              NaN              NaN              NaN   \n",
       "KB-07-13-PRX-07-1              0.0              2.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              2.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              2.0              0.0   \n",
       "\n",
       "Gene ID                                                               \\\n",
       "Gene                                                                   \n",
       "Mutation Name        PM2-Leu277Val    PM2-Leu321Val    PM2-Leu398Val   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Leu277Val        Leu321Val        Leu398Val   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              NaN   \n",
       "KB-07-13-PRX-07-1              0.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              NaN              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                                                PF3D7_1408100  \\\n",
       "Gene                                                             PM3   \n",
       "Mutation Name        PM2-Phe350Val    PM2-Thr289Ile     PM3-Arg31Lys   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Phe350Val        Thr289Ile         Arg31Lys   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              NaN              0.0              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              NaN   \n",
       "KB-07-13-PRX-07-1              2.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              NaN              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                             \n",
       "Gene                                \n",
       "Mutation Name        PM3-Thr375Met  \n",
       "ExonicFunc        missense_variant  \n",
       "AA Change                Thr375Met  \n",
       "Targeted                        No  \n",
       "KB-04-06-PRX-04-1              NaN  \n",
       "AG-00-45-PRX-00-1              0.0  \n",
       "KB-07-13-PRX-07-1              0.0  \n",
       "AM-04-08-PRX-04-1              NaN  \n",
       "AM-07-84-PRX-07-1              0.0  \n",
       "\n",
       "[5 rows x 133 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RUN\n",
    "genotypes = gt_calls[\"genotypes\"]\n",
    "genotypes.to_csv(os.path.join(\n",
    "        wdir, \"filtered_genotypes_table.csv\"))\n",
    "genotypes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5a17d7d",
   "metadata": {},
   "source": [
    "**prevalences**: take the *genotypes* table, if a cell's value is 2, reset its value to 1; otherwise leave as is."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0918fdf7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Gene ID</th>\n",
       "      <th colspan=\"6\" halign=\"left\">PF3D7_0417200</th>\n",
       "      <th colspan=\"4\" halign=\"left\">PF3D7_0523000</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PF3D7_1408000</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PF3D7_1408100</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th colspan=\"6\" halign=\"left\">dhfr-ts</th>\n",
       "      <th colspan=\"4\" halign=\"left\">mdr1</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PM2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PM3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation Name</th>\n",
       "      <th>dhfr-ts-Ala16Val</th>\n",
       "      <th>dhfr-ts-Asn51Ile</th>\n",
       "      <th>dhfr-ts-Cys59Arg</th>\n",
       "      <th>dhfr-ts-Ile164Leu</th>\n",
       "      <th>dhfr-ts-Ser108Asn</th>\n",
       "      <th>dhfr-ts-Ser108Thr</th>\n",
       "      <th>mdr1-Ala821Thr</th>\n",
       "      <th>mdr1-Asn1042Asp</th>\n",
       "      <th>mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>mdr1-Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>PM2-Asp440Asn</th>\n",
       "      <th>PM2-Gln442His</th>\n",
       "      <th>PM2-Ile401Leu</th>\n",
       "      <th>PM2-Leu277Val</th>\n",
       "      <th>PM2-Leu321Val</th>\n",
       "      <th>PM2-Leu398Val</th>\n",
       "      <th>PM2-Phe350Val</th>\n",
       "      <th>PM2-Thr289Ile</th>\n",
       "      <th>PM3-Arg31Lys</th>\n",
       "      <th>PM3-Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ExonicFunc</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>missense_variant&amp;conservative_inframe_deletion</th>\n",
       "      <th>...</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA Change</th>\n",
       "      <th>Ala16Val</th>\n",
       "      <th>Asn51Ile</th>\n",
       "      <th>Cys59Arg</th>\n",
       "      <th>Ile164Leu</th>\n",
       "      <th>Ser108Asn</th>\n",
       "      <th>Ser108Thr</th>\n",
       "      <th>Ala821Thr</th>\n",
       "      <th>Asn1042Asp</th>\n",
       "      <th>Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp</th>\n",
       "      <th>Asn647_Asn654delinsAspAsp</th>\n",
       "      <th>...</th>\n",
       "      <th>Asp440Asn</th>\n",
       "      <th>Gln442His</th>\n",
       "      <th>Ile401Leu</th>\n",
       "      <th>Leu277Val</th>\n",
       "      <th>Leu321Val</th>\n",
       "      <th>Leu398Val</th>\n",
       "      <th>Phe350Val</th>\n",
       "      <th>Thr289Ile</th>\n",
       "      <th>Arg31Lys</th>\n",
       "      <th>Thr375Met</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Targeted</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>...</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KB-04-06-PRX-04-1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-00-45-PRX-00-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KB-07-13-PRX-07-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-04-08-PRX-04-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-07-84-PRX-07-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 133 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene ID              PF3D7_0417200                                    \\\n",
       "Gene                       dhfr-ts                                     \n",
       "Mutation Name     dhfr-ts-Ala16Val dhfr-ts-Asn51Ile dhfr-ts-Cys59Arg   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                 Ala16Val         Asn51Ile         Cys59Arg   \n",
       "Targeted                       Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              1.0              1.0   \n",
       "KB-07-13-PRX-07-1              0.0              1.0              1.0   \n",
       "AM-04-08-PRX-04-1              0.0              1.0              1.0   \n",
       "AM-07-84-PRX-07-1              0.0              1.0              1.0   \n",
       "\n",
       "Gene ID                                                                  \\\n",
       "Gene                                                                      \n",
       "Mutation Name     dhfr-ts-Ile164Leu dhfr-ts-Ser108Asn dhfr-ts-Ser108Thr   \n",
       "ExonicFunc         missense_variant  missense_variant  missense_variant   \n",
       "AA Change                 Ile164Leu         Ser108Asn         Ser108Thr   \n",
       "Targeted                        Yes               Yes               Yes   \n",
       "KB-04-06-PRX-04-1               NaN               NaN               NaN   \n",
       "AG-00-45-PRX-00-1               NaN               1.0               0.0   \n",
       "KB-07-13-PRX-07-1               1.0               1.0               0.0   \n",
       "AM-04-08-PRX-04-1               1.0               1.0               0.0   \n",
       "AM-07-84-PRX-07-1               1.0               1.0               0.0   \n",
       "\n",
       "Gene ID              PF3D7_0523000                   \\\n",
       "Gene                          mdr1                    \n",
       "Mutation Name       mdr1-Ala821Thr  mdr1-Asn1042Asp   \n",
       "ExonicFunc        missense_variant missense_variant   \n",
       "AA Change                Ala821Thr       Asn1042Asp   \n",
       "Targeted                        No              Yes   \n",
       "KB-04-06-PRX-04-1              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              NaN              0.0   \n",
       "KB-07-13-PRX-07-1              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0   \n",
       "\n",
       "Gene ID                                                             \\\n",
       "Gene                                                                 \n",
       "Mutation Name     mdr1-Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "ExonicFunc          missense_variant&conservative_inframe_deletion   \n",
       "AA Change              Asn644_Asn654delinsLysAsnAsnAsnAsnAsnAspAsp   \n",
       "Targeted                                                        No   \n",
       "KB-04-06-PRX-04-1                                              NaN   \n",
       "AG-00-45-PRX-00-1                                              0.0   \n",
       "KB-07-13-PRX-07-1                                              0.0   \n",
       "AM-04-08-PRX-04-1                                              NaN   \n",
       "AM-07-84-PRX-07-1                                              0.0   \n",
       "\n",
       "Gene ID                                                           ...  \\\n",
       "Gene                                                              ...   \n",
       "Mutation Name                     mdr1-Asn647_Asn654delinsAspAsp  ...   \n",
       "ExonicFunc        missense_variant&conservative_inframe_deletion  ...   \n",
       "AA Change                              Asn647_Asn654delinsAspAsp  ...   \n",
       "Targeted                                                      No  ...   \n",
       "KB-04-06-PRX-04-1                                            NaN  ...   \n",
       "AG-00-45-PRX-00-1                                            0.0  ...   \n",
       "KB-07-13-PRX-07-1                                            0.0  ...   \n",
       "AM-04-08-PRX-04-1                                            NaN  ...   \n",
       "AM-07-84-PRX-07-1                                            0.0  ...   \n",
       "\n",
       "Gene ID              PF3D7_1408000                                    \\\n",
       "Gene                           PM2                                     \n",
       "Mutation Name        PM2-Asp440Asn    PM2-Gln442His    PM2-Ile401Leu   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Asp440Asn        Gln442His        Ile401Leu   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              NaN              NaN              NaN   \n",
       "KB-07-13-PRX-07-1              0.0              1.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              1.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              1.0              0.0   \n",
       "\n",
       "Gene ID                                                               \\\n",
       "Gene                                                                   \n",
       "Mutation Name        PM2-Leu277Val    PM2-Leu321Val    PM2-Leu398Val   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Leu277Val        Leu321Val        Leu398Val   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              NaN   \n",
       "KB-07-13-PRX-07-1              0.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              NaN              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                                                PF3D7_1408100  \\\n",
       "Gene                                                             PM3   \n",
       "Mutation Name        PM2-Phe350Val    PM2-Thr289Ile     PM3-Arg31Lys   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Phe350Val        Thr289Ile         Arg31Lys   \n",
       "Targeted                        No               No               No   \n",
       "KB-04-06-PRX-04-1              NaN              0.0              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              NaN   \n",
       "KB-07-13-PRX-07-1              1.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              NaN              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                             \n",
       "Gene                                \n",
       "Mutation Name        PM3-Thr375Met  \n",
       "ExonicFunc        missense_variant  \n",
       "AA Change                Thr375Met  \n",
       "Targeted                        No  \n",
       "KB-04-06-PRX-04-1              NaN  \n",
       "AG-00-45-PRX-00-1              0.0  \n",
       "KB-07-13-PRX-07-1              0.0  \n",
       "AM-04-08-PRX-04-1              NaN  \n",
       "AM-07-84-PRX-07-1              0.0  \n",
       "\n",
       "[5 rows x 133 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RUN\n",
    "prevalences = gt_calls[\"prevalences\"]\n",
    "prevalences.to_csv(os.path.join(wdir, \"prevalences_input_table.csv\"))\n",
    "prevalences.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a25ce785",
   "metadata": {},
   "source": [
    "## Filter genotypes / prevalences\n",
    "It is generally a good idea to do some basic noise removal once the genotypes are created. Some suggestions are provided here."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41c1259e",
   "metadata": {},
   "source": [
    "### Filter variants that are always at low WSAF\n",
    "If a variant is only seen at a low frequency within samples, it is a good indication that it could be just noise. Here we will set a number of samples and minimum WSAF threshold to remove such noise.\n",
    "\n",
    "```python\n",
    "num_samples_wsaf = 2\n",
    "min_wsaf = 0.5\n",
    "wsaf_filter = ((freq > min_wsaf).sum()) >= num_samples_wsaf\n",
    "```\n",
    "\n",
    "The above options will keep the variants that are in at > 0.5 WSAF in at least 2 samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3623718f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# USER INPUT\n",
    "num_samples_wsaf = 2\n",
    "min_wsaf = 0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c7a1f90d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34 of 133 variants will remain after the wsaf filter\n"
     ]
    }
   ],
   "source": [
    "wsaf_filter = ((freq > min_wsaf).sum()) >= num_samples_wsaf\n",
    "print((\"{} of {} variants will remain after the wsaf filter\").format(\n",
    "    wsaf_filter.sum(), freq.shape[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6797b30b",
   "metadata": {},
   "source": [
    "### Filter variants that are observed with low UMI counts\n",
    "If a variant is only supported by a low number of UMIs across the entire sample set, it is another indication of noise.\n",
    "\n",
    "```python\n",
    "num_samples_umi = 2\n",
    "min_umi = 3\n",
    "umi_filter = ((filtered_mutation_counts >= min_umi).sum()) > num_samples_umi\n",
    "```\n",
    "\n",
    "The above options will keep the variants that are supported by at least 3 UMIs in at least 2 samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5d5e53b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# USER INPUT\n",
    "num_samples_umi = 2\n",
    "min_umi = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1887f461",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52 of 133 variants will remain after the UMI filter\n"
     ]
    }
   ],
   "source": [
    "# RUN\n",
    "umi_filter = ((filtered_mutation_counts >= min_umi).sum()) > num_samples_umi\n",
    "print((\"{} of {} variants will remain after the UMI filter\").format(\n",
    "    umi_filter.sum(), freq.shape[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c80589b",
   "metadata": {},
   "source": [
    "### Keep variants that were targeted\n",
    "In most projects there are a number of variants that we would like to report, even if they are not seen in the sample set. We would like to stop those variants from being removed by the above filters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7e39407b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN\n",
    "targ = freq.columns.get_level_values(\"Targeted\") == \"Yes\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11625ca6",
   "metadata": {},
   "source": [
    "### Combine filters\n",
    "Keep the variants that are either targeted or passing filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "183dba3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "79 variants will remain in the final call set.\n",
      "60 variants were targeted and will be kept; and 32 will be removed by the combined UMI and WSAF filters.\n"
     ]
    }
   ],
   "source": [
    "variant_mask = targ | (wsaf_filter & umi_filter)\n",
    "print((\"{} variants will remain in the final call set.\\n\"\n",
    "       \"{} variants were targeted and will be kept; and {} will be removed by \"\n",
    "       \"the combined UMI and WSAF filters.\").format(\n",
    "    variant_mask.sum(), targ.sum(), (wsaf_filter & umi_filter).sum()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be875d3a",
   "metadata": {},
   "source": [
    "## Filter data tables with the combined filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0038335c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Gene ID</th>\n",
       "      <th colspan=\"6\" halign=\"left\">PF3D7_0417200</th>\n",
       "      <th colspan=\"4\" halign=\"left\">PF3D7_0523000</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PF3D7_1343700</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PF3D7_1408000</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th colspan=\"6\" halign=\"left\">dhfr-ts</th>\n",
       "      <th colspan=\"4\" halign=\"left\">mdr1</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">k13</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PM2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation Name</th>\n",
       "      <th>dhfr-ts-Ala16Val</th>\n",
       "      <th>dhfr-ts-Asn51Ile</th>\n",
       "      <th>dhfr-ts-Cys59Arg</th>\n",
       "      <th>dhfr-ts-Ile164Leu</th>\n",
       "      <th>dhfr-ts-Ser108Asn</th>\n",
       "      <th>dhfr-ts-Ser108Thr</th>\n",
       "      <th>mdr1-Asn1042Asp</th>\n",
       "      <th>mdr1-Asn652Asp</th>\n",
       "      <th>mdr1-Asn659_Asn661del</th>\n",
       "      <th>mdr1-Asn660_Asn661del</th>\n",
       "      <th>...</th>\n",
       "      <th>k13-Met476Ile</th>\n",
       "      <th>k13-Phe446Ile</th>\n",
       "      <th>k13-Pro441Leu</th>\n",
       "      <th>k13-Pro527His</th>\n",
       "      <th>k13-Pro553Leu</th>\n",
       "      <th>k13-Pro574Leu</th>\n",
       "      <th>k13-Tyr493His</th>\n",
       "      <th>k13-Val568Gly</th>\n",
       "      <th>PM2-Gln442His</th>\n",
       "      <th>PM2-Leu277Val</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ExonicFunc</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>disruptive_inframe_deletion</th>\n",
       "      <th>disruptive_inframe_deletion</th>\n",
       "      <th>...</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA Change</th>\n",
       "      <th>Ala16Val</th>\n",
       "      <th>Asn51Ile</th>\n",
       "      <th>Cys59Arg</th>\n",
       "      <th>Ile164Leu</th>\n",
       "      <th>Ser108Asn</th>\n",
       "      <th>Ser108Thr</th>\n",
       "      <th>Asn1042Asp</th>\n",
       "      <th>Asn652Asp</th>\n",
       "      <th>Asn659_Asn661del</th>\n",
       "      <th>Asn660_Asn661del</th>\n",
       "      <th>...</th>\n",
       "      <th>Met476Ile</th>\n",
       "      <th>Phe446Ile</th>\n",
       "      <th>Pro441Leu</th>\n",
       "      <th>Pro527His</th>\n",
       "      <th>Pro553Leu</th>\n",
       "      <th>Pro574Leu</th>\n",
       "      <th>Tyr493His</th>\n",
       "      <th>Val568Gly</th>\n",
       "      <th>Gln442His</th>\n",
       "      <th>Leu277Val</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Targeted</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>...</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KB-04-06-PRX-04-1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-00-45-PRX-00-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KB-07-13-PRX-07-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-04-08-PRX-04-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-07-84-PRX-07-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene ID              PF3D7_0417200                                    \\\n",
       "Gene                       dhfr-ts                                     \n",
       "Mutation Name     dhfr-ts-Ala16Val dhfr-ts-Asn51Ile dhfr-ts-Cys59Arg   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                 Ala16Val         Asn51Ile         Cys59Arg   \n",
       "Targeted                       Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              2.0              2.0   \n",
       "KB-07-13-PRX-07-1              0.0              2.0              2.0   \n",
       "AM-04-08-PRX-04-1              0.0              2.0              2.0   \n",
       "AM-07-84-PRX-07-1              0.0              2.0              2.0   \n",
       "\n",
       "Gene ID                                                                  \\\n",
       "Gene                                                                      \n",
       "Mutation Name     dhfr-ts-Ile164Leu dhfr-ts-Ser108Asn dhfr-ts-Ser108Thr   \n",
       "ExonicFunc         missense_variant  missense_variant  missense_variant   \n",
       "AA Change                 Ile164Leu         Ser108Asn         Ser108Thr   \n",
       "Targeted                        Yes               Yes               Yes   \n",
       "KB-04-06-PRX-04-1               NaN               NaN               NaN   \n",
       "AG-00-45-PRX-00-1               NaN               2.0               0.0   \n",
       "KB-07-13-PRX-07-1               2.0               2.0               0.0   \n",
       "AM-04-08-PRX-04-1               1.0               2.0               0.0   \n",
       "AM-07-84-PRX-07-1               2.0               2.0               0.0   \n",
       "\n",
       "Gene ID              PF3D7_0523000                   \\\n",
       "Gene                          mdr1                    \n",
       "Mutation Name      mdr1-Asn1042Asp   mdr1-Asn652Asp   \n",
       "ExonicFunc        missense_variant missense_variant   \n",
       "AA Change               Asn1042Asp        Asn652Asp   \n",
       "Targeted                       Yes               No   \n",
       "KB-04-06-PRX-04-1              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              0.0   \n",
       "KB-07-13-PRX-07-1              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              NaN   \n",
       "AM-07-84-PRX-07-1              0.0              0.0   \n",
       "\n",
       "Gene ID                                                                    \\\n",
       "Gene                                                                        \n",
       "Mutation Name           mdr1-Asn659_Asn661del       mdr1-Asn660_Asn661del   \n",
       "ExonicFunc        disruptive_inframe_deletion disruptive_inframe_deletion   \n",
       "AA Change                    Asn659_Asn661del            Asn660_Asn661del   \n",
       "Targeted                                   No                          No   \n",
       "KB-04-06-PRX-04-1                         NaN                         NaN   \n",
       "AG-00-45-PRX-00-1                         0.0                         0.0   \n",
       "KB-07-13-PRX-07-1                         0.0                         0.0   \n",
       "AM-04-08-PRX-04-1                         NaN                         NaN   \n",
       "AM-07-84-PRX-07-1                         0.0                         0.0   \n",
       "\n",
       "Gene ID            ...    PF3D7_1343700                                    \\\n",
       "Gene               ...              k13                                     \n",
       "Mutation Name      ...    k13-Met476Ile    k13-Phe446Ile    k13-Pro441Leu   \n",
       "ExonicFunc         ... missense_variant missense_variant missense_variant   \n",
       "AA Change          ...        Met476Ile        Phe446Ile        Pro441Leu   \n",
       "Targeted           ...              Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1  ...              0.0              NaN              NaN   \n",
       "AG-00-45-PRX-00-1  ...              0.0              0.0              0.0   \n",
       "KB-07-13-PRX-07-1  ...              0.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1  ...              0.0              0.0              0.0   \n",
       "AM-07-84-PRX-07-1  ...              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                                                               \\\n",
       "Gene                                                                   \n",
       "Mutation Name        k13-Pro527His    k13-Pro553Leu    k13-Pro574Leu   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Pro527His        Pro553Leu        Pro574Leu   \n",
       "Targeted                       Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              0.0   \n",
       "KB-07-13-PRX-07-1              0.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                                                PF3D7_1408000  \\\n",
       "Gene                                                             PM2   \n",
       "Mutation Name        k13-Tyr493His    k13-Val568Gly    PM2-Gln442His   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Tyr493His        Val568Gly        Gln442His   \n",
       "Targeted                       Yes              Yes               No   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              NaN   \n",
       "KB-07-13-PRX-07-1              0.0              0.0              2.0   \n",
       "AM-04-08-PRX-04-1              0.0              0.0              2.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              2.0   \n",
       "\n",
       "Gene ID                             \n",
       "Gene                                \n",
       "Mutation Name        PM2-Leu277Val  \n",
       "ExonicFunc        missense_variant  \n",
       "AA Change                Leu277Val  \n",
       "Targeted                        No  \n",
       "KB-04-06-PRX-04-1              NaN  \n",
       "AG-00-45-PRX-00-1              0.0  \n",
       "KB-07-13-PRX-07-1              0.0  \n",
       "AM-04-08-PRX-04-1              NaN  \n",
       "AM-07-84-PRX-07-1              0.0  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_genotypes = genotypes.loc[:, variant_mask]\n",
    "filtered_genotypes.to_csv(os.path.join(wdir, \"final_filtered_genotypes.csv\"))\n",
    "filtered_genotypes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4586302f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Gene ID</th>\n",
       "      <th colspan=\"6\" halign=\"left\">PF3D7_0417200</th>\n",
       "      <th colspan=\"4\" halign=\"left\">PF3D7_0523000</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">PF3D7_1343700</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PF3D7_1408000</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th colspan=\"6\" halign=\"left\">dhfr-ts</th>\n",
       "      <th colspan=\"4\" halign=\"left\">mdr1</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">k13</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PM2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation Name</th>\n",
       "      <th>dhfr-ts-Ala16Val</th>\n",
       "      <th>dhfr-ts-Asn51Ile</th>\n",
       "      <th>dhfr-ts-Cys59Arg</th>\n",
       "      <th>dhfr-ts-Ile164Leu</th>\n",
       "      <th>dhfr-ts-Ser108Asn</th>\n",
       "      <th>dhfr-ts-Ser108Thr</th>\n",
       "      <th>mdr1-Asn1042Asp</th>\n",
       "      <th>mdr1-Asn652Asp</th>\n",
       "      <th>mdr1-Asn659_Asn661del</th>\n",
       "      <th>mdr1-Asn660_Asn661del</th>\n",
       "      <th>...</th>\n",
       "      <th>k13-Met476Ile</th>\n",
       "      <th>k13-Phe446Ile</th>\n",
       "      <th>k13-Pro441Leu</th>\n",
       "      <th>k13-Pro527His</th>\n",
       "      <th>k13-Pro553Leu</th>\n",
       "      <th>k13-Pro574Leu</th>\n",
       "      <th>k13-Tyr493His</th>\n",
       "      <th>k13-Val568Gly</th>\n",
       "      <th>PM2-Gln442His</th>\n",
       "      <th>PM2-Leu277Val</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ExonicFunc</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>disruptive_inframe_deletion</th>\n",
       "      <th>disruptive_inframe_deletion</th>\n",
       "      <th>...</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA Change</th>\n",
       "      <th>Ala16Val</th>\n",
       "      <th>Asn51Ile</th>\n",
       "      <th>Cys59Arg</th>\n",
       "      <th>Ile164Leu</th>\n",
       "      <th>Ser108Asn</th>\n",
       "      <th>Ser108Thr</th>\n",
       "      <th>Asn1042Asp</th>\n",
       "      <th>Asn652Asp</th>\n",
       "      <th>Asn659_Asn661del</th>\n",
       "      <th>Asn660_Asn661del</th>\n",
       "      <th>...</th>\n",
       "      <th>Met476Ile</th>\n",
       "      <th>Phe446Ile</th>\n",
       "      <th>Pro441Leu</th>\n",
       "      <th>Pro527His</th>\n",
       "      <th>Pro553Leu</th>\n",
       "      <th>Pro574Leu</th>\n",
       "      <th>Tyr493His</th>\n",
       "      <th>Val568Gly</th>\n",
       "      <th>Gln442His</th>\n",
       "      <th>Leu277Val</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Targeted</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "      <th>...</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>Yes</th>\n",
       "      <th>No</th>\n",
       "      <th>No</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KB-04-06-PRX-04-1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG-00-45-PRX-00-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KB-07-13-PRX-07-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-04-08-PRX-04-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-07-84-PRX-07-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene ID              PF3D7_0417200                                    \\\n",
       "Gene                       dhfr-ts                                     \n",
       "Mutation Name     dhfr-ts-Ala16Val dhfr-ts-Asn51Ile dhfr-ts-Cys59Arg   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                 Ala16Val         Asn51Ile         Cys59Arg   \n",
       "Targeted                       Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              1.0              1.0   \n",
       "KB-07-13-PRX-07-1              0.0              1.0              1.0   \n",
       "AM-04-08-PRX-04-1              0.0              1.0              1.0   \n",
       "AM-07-84-PRX-07-1              0.0              1.0              1.0   \n",
       "\n",
       "Gene ID                                                                  \\\n",
       "Gene                                                                      \n",
       "Mutation Name     dhfr-ts-Ile164Leu dhfr-ts-Ser108Asn dhfr-ts-Ser108Thr   \n",
       "ExonicFunc         missense_variant  missense_variant  missense_variant   \n",
       "AA Change                 Ile164Leu         Ser108Asn         Ser108Thr   \n",
       "Targeted                        Yes               Yes               Yes   \n",
       "KB-04-06-PRX-04-1               NaN               NaN               NaN   \n",
       "AG-00-45-PRX-00-1               NaN               1.0               0.0   \n",
       "KB-07-13-PRX-07-1               1.0               1.0               0.0   \n",
       "AM-04-08-PRX-04-1               1.0               1.0               0.0   \n",
       "AM-07-84-PRX-07-1               1.0               1.0               0.0   \n",
       "\n",
       "Gene ID              PF3D7_0523000                   \\\n",
       "Gene                          mdr1                    \n",
       "Mutation Name      mdr1-Asn1042Asp   mdr1-Asn652Asp   \n",
       "ExonicFunc        missense_variant missense_variant   \n",
       "AA Change               Asn1042Asp        Asn652Asp   \n",
       "Targeted                       Yes               No   \n",
       "KB-04-06-PRX-04-1              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              0.0   \n",
       "KB-07-13-PRX-07-1              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              NaN   \n",
       "AM-07-84-PRX-07-1              0.0              0.0   \n",
       "\n",
       "Gene ID                                                                    \\\n",
       "Gene                                                                        \n",
       "Mutation Name           mdr1-Asn659_Asn661del       mdr1-Asn660_Asn661del   \n",
       "ExonicFunc        disruptive_inframe_deletion disruptive_inframe_deletion   \n",
       "AA Change                    Asn659_Asn661del            Asn660_Asn661del   \n",
       "Targeted                                   No                          No   \n",
       "KB-04-06-PRX-04-1                         NaN                         NaN   \n",
       "AG-00-45-PRX-00-1                         0.0                         0.0   \n",
       "KB-07-13-PRX-07-1                         0.0                         0.0   \n",
       "AM-04-08-PRX-04-1                         NaN                         NaN   \n",
       "AM-07-84-PRX-07-1                         0.0                         0.0   \n",
       "\n",
       "Gene ID            ...    PF3D7_1343700                                    \\\n",
       "Gene               ...              k13                                     \n",
       "Mutation Name      ...    k13-Met476Ile    k13-Phe446Ile    k13-Pro441Leu   \n",
       "ExonicFunc         ... missense_variant missense_variant missense_variant   \n",
       "AA Change          ...        Met476Ile        Phe446Ile        Pro441Leu   \n",
       "Targeted           ...              Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1  ...              0.0              NaN              NaN   \n",
       "AG-00-45-PRX-00-1  ...              0.0              0.0              0.0   \n",
       "KB-07-13-PRX-07-1  ...              0.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1  ...              0.0              0.0              0.0   \n",
       "AM-07-84-PRX-07-1  ...              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                                                               \\\n",
       "Gene                                                                   \n",
       "Mutation Name        k13-Pro527His    k13-Pro553Leu    k13-Pro574Leu   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Pro527His        Pro553Leu        Pro574Leu   \n",
       "Targeted                       Yes              Yes              Yes   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              0.0   \n",
       "KB-07-13-PRX-07-1              0.0              0.0              0.0   \n",
       "AM-04-08-PRX-04-1              0.0              0.0              0.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              0.0   \n",
       "\n",
       "Gene ID                                                PF3D7_1408000  \\\n",
       "Gene                                                             PM2   \n",
       "Mutation Name        k13-Tyr493His    k13-Val568Gly    PM2-Gln442His   \n",
       "ExonicFunc        missense_variant missense_variant missense_variant   \n",
       "AA Change                Tyr493His        Val568Gly        Gln442His   \n",
       "Targeted                       Yes              Yes               No   \n",
       "KB-04-06-PRX-04-1              NaN              NaN              NaN   \n",
       "AG-00-45-PRX-00-1              0.0              0.0              NaN   \n",
       "KB-07-13-PRX-07-1              0.0              0.0              1.0   \n",
       "AM-04-08-PRX-04-1              0.0              0.0              1.0   \n",
       "AM-07-84-PRX-07-1              0.0              0.0              1.0   \n",
       "\n",
       "Gene ID                             \n",
       "Gene                                \n",
       "Mutation Name        PM2-Leu277Val  \n",
       "ExonicFunc        missense_variant  \n",
       "AA Change                Leu277Val  \n",
       "Targeted                        No  \n",
       "KB-04-06-PRX-04-1              NaN  \n",
       "AG-00-45-PRX-00-1              0.0  \n",
       "KB-07-13-PRX-07-1              0.0  \n",
       "AM-04-08-PRX-04-1              NaN  \n",
       "AM-07-84-PRX-07-1              0.0  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_prevalences = prevalences.loc[:, variant_mask]\n",
    "filtered_prevalences.to_csv(os.path.join(wdir, \"final_filtered_prevalences_input_table.csv\"))\n",
    "filtered_prevalences.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "153c029e",
   "metadata": {},
   "source": [
    "# Calculate Prevalence in Each Region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "487145e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_file = '/opt/prevalence_metadata/PRX-00_metadata.csv'\n",
    "prevalences_input_table = 'prevalences_input_table.csv'\n",
    "UMI_suffix = '-PRX-00-1'\n",
    "output_summary_table = 'prevalence_summary.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "45f4b571-2402-4766-a1cf-5baecd7879e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations = [\n",
    "\"crt-Cys72Ser\",\n",
    "\"crt-Val73Leu\",\n",
    "\"crt-Met74Ile\",\n",
    "\"crt-Asn75Glu\",\n",
    "\"crt-Lys76Thr\",\n",
    "\"crt-Asn326Ser\",\n",
    "\"crt-Ile356Thr\",\n",
    "\"dhfr-ts-Asn51Ile\",\n",
    "\"dhfr-ts-Cys59Arg\",\n",
    "\"dhfr-ts-Ser108Asn\",\n",
    "\"dhfr-ts-Ser108Thr\",\n",
    "\"dhfr-ts-Ile164Leu\",\n",
    "\"dhps-Ala437Gly\",\n",
    "\"dhps-Ala581Gly\",\n",
    "\"dhps-Ala613Ser\",\n",
    "\"dhps-Ala613Thr\",\n",
    "\"dhps-Ile431Val\",\n",
    "\"dhps-Lys540Glu\",\n",
    "\"dhps-Ser436Ala\",\n",
    "\"dhps-Ser436Phe\",\n",
    "\"k13-Ala675Val\",\n",
    "\"k13-Arg539Thr\",\n",
    "\"k13-Arg561His\",\n",
    "\"k13-Arg622Ile\",\n",
    "\"k13-Cys469Phe\",\n",
    "\"k13-Cys469Tyr\",\n",
    "\"k13-Cys580Tyr\",\n",
    "\"k13-Pro441Leu\",\n",
    "\"k13-Tyr493His\",\n",
    "\"k13-Val568Gly\",\n",
    "\"mdr1-Asn1042Asp\",\n",
    "\"mdr1-Asn86Phe\",\n",
    "\"mdr1-Asn86Tyr\",\n",
    "\"mdr1-Asp1246Tyr\",\n",
    "\"mdr1-Ser1034Cys\",\n",
    "\"mdr1-Tyr184Phe\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d0c5ba97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dhfr-ts-Asn51Ile</th>\n",
       "      <th>dhfr-ts-Cys59Arg</th>\n",
       "      <th>dhfr-ts-Ile164Leu</th>\n",
       "      <th>dhfr-ts-Ser108Asn</th>\n",
       "      <th>dhfr-ts-Ser108Thr</th>\n",
       "      <th>mdr1-Asn1042Asp</th>\n",
       "      <th>mdr1-Asn86Tyr</th>\n",
       "      <th>mdr1-Asp1246Tyr</th>\n",
       "      <th>mdr1-Ser1034Cys</th>\n",
       "      <th>mdr1-Tyr184Phe</th>\n",
       "      <th>...</th>\n",
       "      <th>k13-Ala675Val</th>\n",
       "      <th>k13-Arg539Thr</th>\n",
       "      <th>k13-Arg561His</th>\n",
       "      <th>k13-Arg622Ile</th>\n",
       "      <th>k13-Cys469Phe</th>\n",
       "      <th>k13-Cys469Tyr</th>\n",
       "      <th>k13-Cys580Tyr</th>\n",
       "      <th>k13-Pro441Leu</th>\n",
       "      <th>k13-Tyr493His</th>\n",
       "      <th>k13-Val568Gly</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sites</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Agago</th>\n",
       "      <td>1.0 (18/18)</td>\n",
       "      <td>0.7777777777777778 (14/18)</td>\n",
       "      <td>0.07142857142857142 (1/14)</td>\n",
       "      <td>1.0 (16/16)</td>\n",
       "      <td>0 (0/16)</td>\n",
       "      <td>0 (0/20)</td>\n",
       "      <td>0 (0/18)</td>\n",
       "      <td>0.14285714285714285 (2/14)</td>\n",
       "      <td>0 (0/20)</td>\n",
       "      <td>0.65 (13/20)</td>\n",
       "      <td>...</td>\n",
       "      <td>0 (0/20)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/15)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/16)</td>\n",
       "      <td>0 (0/15)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Amolatar</th>\n",
       "      <td>1.0 (13/13)</td>\n",
       "      <td>1.0 (13/13)</td>\n",
       "      <td>0 (0/8)</td>\n",
       "      <td>1.0 (8/8)</td>\n",
       "      <td>0 (0/8)</td>\n",
       "      <td>0 (0/18)</td>\n",
       "      <td>0 (0/14)</td>\n",
       "      <td>0 (0/6)</td>\n",
       "      <td>0 (0/18)</td>\n",
       "      <td>1.0 (14/14)</td>\n",
       "      <td>...</td>\n",
       "      <td>0 (0/18)</td>\n",
       "      <td>0 (0/18)</td>\n",
       "      <td>0 (0/9)</td>\n",
       "      <td>0 (0/18)</td>\n",
       "      <td>0 (0/16)</td>\n",
       "      <td>0 (0/16)</td>\n",
       "      <td>0 (0/18)</td>\n",
       "      <td>0 (0/16)</td>\n",
       "      <td>0 (0/14)</td>\n",
       "      <td>0 (0/9)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Kabale</th>\n",
       "      <td>1.0 (19/19)</td>\n",
       "      <td>1.0 (19/19)</td>\n",
       "      <td>0.23529411764705882 (4/17)</td>\n",
       "      <td>1.0 (16/16)</td>\n",
       "      <td>0 (0/16)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/16)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0.42105263157894735 (8/19)</td>\n",
       "      <td>...</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/17)</td>\n",
       "      <td>0 (0/20)</td>\n",
       "      <td>0.21052631578947367 (4/19)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/20)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/15)</td>\n",
       "      <td>0 (0/17)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Kole</th>\n",
       "      <td>1.0 (11/11)</td>\n",
       "      <td>0.9090909090909091 (10/11)</td>\n",
       "      <td>0 (0/7)</td>\n",
       "      <td>1.0 (7/7)</td>\n",
       "      <td>0 (0/7)</td>\n",
       "      <td>0 (0/17)</td>\n",
       "      <td>0 (0/15)</td>\n",
       "      <td>0 (0/11)</td>\n",
       "      <td>0 (0/17)</td>\n",
       "      <td>0.6153846153846154 (8/13)</td>\n",
       "      <td>...</td>\n",
       "      <td>0 (0/18)</td>\n",
       "      <td>0 (0/19)</td>\n",
       "      <td>0 (0/15)</td>\n",
       "      <td>0 (0/18)</td>\n",
       "      <td>0 (0/16)</td>\n",
       "      <td>0 (0/16)</td>\n",
       "      <td>0 (0/18)</td>\n",
       "      <td>0 (0/16)</td>\n",
       "      <td>0 (0/16)</td>\n",
       "      <td>0 (0/15)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>overall</th>\n",
       "      <td>1.0 (61/61)</td>\n",
       "      <td>0.9180327868852459 (56/61)</td>\n",
       "      <td>0.10869565217391304 (5/46)</td>\n",
       "      <td>1.0 (47/47)</td>\n",
       "      <td>0 (0/47)</td>\n",
       "      <td>0 (0/74)</td>\n",
       "      <td>0 (0/66)</td>\n",
       "      <td>0.0425531914893617 (2/47)</td>\n",
       "      <td>0 (0/74)</td>\n",
       "      <td>0.6515151515151515 (43/66)</td>\n",
       "      <td>...</td>\n",
       "      <td>0 (0/75)</td>\n",
       "      <td>0 (0/75)</td>\n",
       "      <td>0 (0/56)</td>\n",
       "      <td>0 (0/75)</td>\n",
       "      <td>0.05714285714285714 (4/70)</td>\n",
       "      <td>0 (0/70)</td>\n",
       "      <td>0 (0/75)</td>\n",
       "      <td>0 (0/70)</td>\n",
       "      <td>0 (0/61)</td>\n",
       "      <td>0 (0/56)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         dhfr-ts-Asn51Ile            dhfr-ts-Cys59Arg  \\\n",
       "Sites                                                   \n",
       "Agago         1.0 (18/18)  0.7777777777777778 (14/18)   \n",
       "Amolatar      1.0 (13/13)                 1.0 (13/13)   \n",
       "Kabale        1.0 (19/19)                 1.0 (19/19)   \n",
       "Kole          1.0 (11/11)  0.9090909090909091 (10/11)   \n",
       "overall       1.0 (61/61)  0.9180327868852459 (56/61)   \n",
       "\n",
       "                   dhfr-ts-Ile164Leu dhfr-ts-Ser108Asn dhfr-ts-Ser108Thr  \\\n",
       "Sites                                                                      \n",
       "Agago     0.07142857142857142 (1/14)       1.0 (16/16)          0 (0/16)   \n",
       "Amolatar                     0 (0/8)         1.0 (8/8)           0 (0/8)   \n",
       "Kabale    0.23529411764705882 (4/17)       1.0 (16/16)          0 (0/16)   \n",
       "Kole                         0 (0/7)         1.0 (7/7)           0 (0/7)   \n",
       "overall   0.10869565217391304 (5/46)       1.0 (47/47)          0 (0/47)   \n",
       "\n",
       "         mdr1-Asn1042Asp mdr1-Asn86Tyr             mdr1-Asp1246Tyr  \\\n",
       "Sites                                                                \n",
       "Agago           0 (0/20)      0 (0/18)  0.14285714285714285 (2/14)   \n",
       "Amolatar        0 (0/18)      0 (0/14)                     0 (0/6)   \n",
       "Kabale          0 (0/19)      0 (0/19)                    0 (0/16)   \n",
       "Kole            0 (0/17)      0 (0/15)                    0 (0/11)   \n",
       "overall         0 (0/74)      0 (0/66)   0.0425531914893617 (2/47)   \n",
       "\n",
       "         mdr1-Ser1034Cys              mdr1-Tyr184Phe  ... k13-Ala675Val  \\\n",
       "Sites                                                 ...                 \n",
       "Agago           0 (0/20)                0.65 (13/20)  ...      0 (0/20)   \n",
       "Amolatar        0 (0/18)                 1.0 (14/14)  ...      0 (0/18)   \n",
       "Kabale          0 (0/19)  0.42105263157894735 (8/19)  ...      0 (0/19)   \n",
       "Kole            0 (0/17)   0.6153846153846154 (8/13)  ...      0 (0/18)   \n",
       "overall         0 (0/74)  0.6515151515151515 (43/66)  ...      0 (0/75)   \n",
       "\n",
       "         k13-Arg539Thr k13-Arg561His k13-Arg622Ile  \\\n",
       "Sites                                                \n",
       "Agago         0 (0/19)      0 (0/15)      0 (0/19)   \n",
       "Amolatar      0 (0/18)       0 (0/9)      0 (0/18)   \n",
       "Kabale        0 (0/19)      0 (0/17)      0 (0/20)   \n",
       "Kole          0 (0/19)      0 (0/15)      0 (0/18)   \n",
       "overall       0 (0/75)      0 (0/56)      0 (0/75)   \n",
       "\n",
       "                       k13-Cys469Phe k13-Cys469Tyr k13-Cys580Tyr  \\\n",
       "Sites                                                              \n",
       "Agago                       0 (0/19)      0 (0/19)      0 (0/19)   \n",
       "Amolatar                    0 (0/16)      0 (0/16)      0 (0/18)   \n",
       "Kabale    0.21052631578947367 (4/19)      0 (0/19)      0 (0/20)   \n",
       "Kole                        0 (0/16)      0 (0/16)      0 (0/18)   \n",
       "overall   0.05714285714285714 (4/70)      0 (0/70)      0 (0/75)   \n",
       "\n",
       "         k13-Pro441Leu k13-Tyr493His k13-Val568Gly  \n",
       "Sites                                               \n",
       "Agago         0 (0/19)      0 (0/16)      0 (0/15)  \n",
       "Amolatar      0 (0/16)      0 (0/14)       0 (0/9)  \n",
       "Kabale        0 (0/19)      0 (0/15)      0 (0/17)  \n",
       "Kole          0 (0/16)      0 (0/16)      0 (0/15)  \n",
       "overall       0 (0/70)      0 (0/61)      0 (0/56)  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cap.calculate_prevalences(metadata_file,\n",
    "                          prevalences_input_table,\n",
    "                          UMI_suffix,\n",
    "                          mutations,\n",
    "                          output_summary_table)\n",
    "\n",
    "prevalences_output = pd.read_csv(output_summary_table,sep='\\t', index_col=0)\n",
    "prevalences_output.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83a1a388",
   "metadata": {},
   "source": [
    "# Plot Choropleth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "080505ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e7f75640a8a24a63bbf028671d292ff8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dropdown(description='region:', options=('uganda', 'tanzania'), value='uganda')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "########### USER INPUT #################\n",
    "region_of_interest = 'uganda'\n",
    "w = widgets.Dropdown(\n",
    "    options=['uganda', 'tanzania'],\n",
    "    value='uganda',\n",
    "    description='region:',\n",
    "    disabled=False,\n",
    ")\n",
    "display(w)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "dcf74982-7f91-4258-9e9b-dd6e9efdc9f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "54ed6e01ed8c4a269bf3b1cc4b696083",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='variant', options=('crt-Asn326Ser', 'crt-Asn75Glu', 'crt-Cys72Ser'…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#RUN\n",
    "with urlopen('https://raw.githubusercontent.com/simkin-bioinformatics/geojson_files/main/uganda-with-regions.geojson') as response:\n",
    "    uganda_regions = json.load(response)\n",
    "with urlopen('https://raw.githubusercontent.com/simkin-bioinformatics/geojson_files/main/tanzania-with-regions.geojson') as response:\n",
    "    tanzania_regions = json.load(response)\n",
    "\n",
    "\n",
    "# load the inputs\n",
    "region_dict = {'uganda':[uganda_regions,5.8, {\"lat\": 1.3733, \"lon\": 32.2903}],\n",
    "               'tanzania':[tanzania_regions,4, {\"lat\": -6.3690, \"lon\": 34.8888}]}\n",
    "\n",
    "prevalence_summary_df = pd.read_csv(output_summary_table, sep='\\t')\n",
    "\n",
    "# identify columns that are variants not headers\n",
    "variant_columns=prevalence_summary_df.columns.difference(['Sites'])\n",
    "\n",
    "# get a list of every variant from the dataframe and convert prevalence values to floats\n",
    "variant_list = []\n",
    "for column in variant_columns:\n",
    "    prevalence_summary_df[column]=[float(x.split()[0]) for x in prevalence_summary_df[column]]\n",
    "    variant_list.append(column)\n",
    "\n",
    "def display_choropleth(variant):\n",
    "    json_file = region_dict[w.value][0]\n",
    "    fig = px.choropleth_mapbox(prevalence_summary_df, \n",
    "                                geojson=json_file, \n",
    "                                locations='Sites', \n",
    "                                color=variant,\n",
    "                                color_continuous_scale=\"reds\",\n",
    "                                mapbox_style=\"open-street-map\",\n",
    "                                featureidkey=\"properties.name\",\n",
    "                                zoom=region_dict[w.value][1], \n",
    "                                center = region_dict[w.value][2],\n",
    "                                labels={'Sites':'Site'},\n",
    "                                range_color=(0,1),\n",
    "    )\n",
    "    fig.update_layout(margin={\"r\":0,\"t\":40,\"l\":0,\"b\":0})\n",
    "    fig.update_layout(height=500)\n",
    "    return fig\n",
    "\n",
    "interact(display_choropleth, variant=variant_list);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d02dde81-54c5-4133-82aa-1d36a6f9c2c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook prevalence_plotting.ipynb to html\n",
      "[NbConvertApp] Writing 730157 bytes to prevalence_plotting.html\n"
     ]
    }
   ],
   "source": [
    "!jupyter nbconvert --to html prevalence_plotting.ipynb"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
